rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md

```diff
@@ -0,0 +1,229 @@
+
+**Immutable Worker Behavior:**
+- Workers receive their complete configuration at startup
+- No runtime configuration changes
+- Worker role and idle timeout never change during worker lifetime
+- `minCapacity` requirements fulfilled through worker replacement, not reconfiguration
+
+#### 3. Pool Configuration
+
+**New Configuration Options:**
+Pools will have a new `minCapacityIdleTimeoutSecs` parameter that enables `minCapacity` worker behavior. Setting this to 0 means `minCapacity` workers will run indefinitely (no idle timeout), providing maximum cache preservation.
+
+**Validation:**
+- Fail pool provisioning entirely if invalid configuration (e.g., `minCapacity > maxCapacity`)
+- Require `minCapacityIdleTimeoutSecs >= 0` if `minCapacity` workers are desired
+- Setting `minCapacityIdleTimeoutSecs = 0` enables indefinite runtime for maximum cache preservation
+
+#### 4. Enhanced Provisioning Logic
+
+**Provisioning Strategy:**
+The enhanced estimator uses existing `minCapacity` enforcement logic but provisions workers with different idle timeout configurations based on pool needs. When total running capacity falls below minCapacity, new workers are spawned with indefinite runtime (minCapacityIdleTimeoutSecs = 0) to serve as long-lived `minCapacity` workers. Additional workers beyond `minCapacity` are spawned with standard idle timeouts. This demand-based approach avoids over-provisioning and works within existing estimator logic.
+
+**Current Estimator Implementation:** [Estimator.simple() method](https://github.com/taskcluster/taskcluster/blob/d7fbf0ee9e0d93e079cc7ff069eaceecfd7d29ec/services/worker-manager/src/estimator.js#L7-L84) enforces `minCapacity` as a floor at [line 35-39](https://github.com/taskcluster/taskcluster/blob/d7fbf0ee9e0d93e079cc7ff069eaceecfd7d29ec/services/worker-manager/src/estimator.js#L35-L39)
+
+**Capacity Management:**
+- **minCapacity increases**: Spawn additional `minCapacity` workers with indefinite runtime
+- **minCapacity decreases**: Forcefully terminate excess `minCapacity` workers (see Capacity Reduction Trade-offs)
+- **Worker failures**: Failed workers are handled through normal provisioning cycles when total capacity falls below minCapacity
+- **Multi-region**: Pool-wide capacity management (any region can fulfill minCapacity)
+
+**Capacity Reduction Trade-offs:**
+When `minCapacity` decreases, excess `minCapacity` workers must be forcefully terminated since:
+- Worker-manager cannot communicate with running workers to reduce their idle timeouts
+- Waiting for natural idle timeout could take hours (defeating responsive capacity management)
+- Forceful termination provides immediate capacity adjustment but loses cache benefits
```

Comment:
whenever we kill a running worker there is a high chance of getting claim expired, as he might be picking up new tasks at the same time. I wonder if we should treat those in some special way. I.e. mark them as "quarantined for removal" to make sure they won't pick new tasks, and only then remove (maybe on the next scanner run)

Reply:
Great call! I changed the RFC to first quarantine the worker.